### ISSUE TITLE
Document score is not accessible in aggregation nested filters

### ISSUE THREAD

**Issue by MauDagos** [elastic/elasticsearch#48791]
**Elasticsearch version** (`bin/elasticsearch --version`):

```json
  "version" : {
    "number" : "5.6.15",
    "build_hash" : "fe7575a",
    "build_date" : "2019-02-13T16:21:45.880Z",
    "build_snapshot" : false,
    "lucene_version" : "6.6.1"
  }
```

**Plugins installed**:

```json
[
  {
    "name": "discovery-ec2",
    "version": "5.6.15",
    "description": "The EC2 discovery plugin allows to use AWS API for the unicast discovery mechanism.",
    "classname": "org.elasticsearch.discovery.ec2.Ec2DiscoveryPlugin",
    "has_native_controller": false
  },
  {
    "name": "mapper-murmur3",
    "version": "5.6.15",
    "description": "The Mapper Murmur3 plugin allows to compute hashes of a field's values at index-time and to store them in the index.",
    "classname": "org.elasticsearch.plugin.mapper.MapperMurmur3Plugin",
    "has_native_controller": false
  },
  {
    "name": "repository-s3",
    "version": "5.6.15",
    "description": "The S3 repository plugin adds S3 repositories",
    "classname": "org.elasticsearch.repositories.s3.S3RepositoryPlugin",
    "has_native_controller": false
  }
]
```

**JVM version** (`java -version`):

```text
java version "1.8.0_161"
Java(TM) SE Runtime Environment (build 1.8.0_161-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.161-b12, mixed mode)
```

**OS version** (`uname -a` if on a Unix-like system):

```text
Linux i-023ce3162e7cea2fe 4.14.101-75.76.amzn1.x86_64 #1 SMP Tue Feb 19 18:14:51 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

--------------

Hello :wave: I am currently exploring the [function score query feature](https://www.elastic.co/guide/en/elasticsearch/reference/5.6/query-dsl-function-score-query.html) and I've run into an unexpected wall.

My index has documents with two nested fields. With `function_score`, we score documents based on whether one of the nested fields passes some filters. Later, we aggregate on the **other** nested field, but we need to know what `_score` the document had (so as to know what filters matched the first nested field).

Here's how our aggregations look like:

```json
{
  "aggs": {
    "agg_01": {
      "nested": {"path": "nested_field_02"},
      "filter": {"bool": {"must": [
        {"exists": {"field": "nested_field_02.my_field"}},
        { "script": {
          "script": {
            "lang": "expression",
            "inline": "_score == 1"
          }
        }}
      ]}}
    }
  }
}
```

An aggregation like the above yields no errors, but the `script` always evaluates to `false`. I have also tried with `"lang": "painless"`, but that does yield an error.

NOTE: Even though I work with ES 5.6, we have also tried this in ES 7.4 with no success.

Let me know if you need any more information. Best regards,

\- Mauro

**Comment by chenyiping111** [comment]
{
  "_source": {
    "includes": [
      "id",
      "_type"
    ]
  },
  "aggs": {
    "authorsId": {
      "aggs": {
        "basci_score": {
          "sum": {
            "script": {
              "source": "if (doc.containsKey('n_citation_ln') ) {return ( 1 + doc['n_citation_ln'].value) * _score}else{return 10*_score}"
            }
          }
        }
      },
      "terms": {
        "field": "authors.id",
        "order": [
          {
            "basci_score": "desc"
          }
        ],
        "size": 1000
      }
    }
  },
  "from": 0,
  "query": {
    "bool": {
      "must": {
        "bool": {
          "must": [
            {
              "query_string": {
                "default_operator": "AND",
                "fields": [
                  "title^10000",
                  "title_zh^10000",
                  "authors.name",
                  "vnames",
                  "publisher",
                  "keywords^20",
                  "keywords_zh^20",
                  "citation^50000"
                ],
                "query": "(douglass c north) OR (douglass c north)"
              }
            }
          ]
        }
      }
    }
  },
  "size": 0
}

my search above, but it's return "reason": "class_cast_exception: Cannot cast from [java.lang.Number] to [int]."  es6.4 don't have this problem

**Comment by MauDagos** [comment]
> {
> "_source": {
> "includes": [
> "id",
> "_type"
> ]
> },
> "aggs": {
> "authorsId": {
> "aggs": {
> "basci_score": {
> "sum": {
> "script": {
> "source": "if (doc.containsKey('n_citation_ln') ) {return ( 1 + doc['n_citation_ln'].value) * _score}else{return 10*_score}"
> }
> }
> }
> },
> "terms": {
> "field": "authors.id",
> "order": [
> {
> "basci_score": "desc"
> }
> ],
> "size": 1000
> }
> }
> },
> "from": 0,
> "query": {
> "bool": {
> "must": {
> "bool": {
> "must": [
> {
> "query_string": {
> "default_operator": "AND",
> "fields": [
> "title^10000",
> "title_zh^10000",
> "authors.name",
> "vnames",
> "publisher",
> "keywords^20",
> "keywords_zh^20",
> "citation^50000"
> ],
> "query": "(douglass c north) OR (douglass c north)"
> }
> }
> ]
> }
> }
> }
> },
> "size": 0
> }
> 
> my search above, but it's return "reason": "class_cast_exception: Cannot cast from [java.lang.Number] to [int]." es6.4 don't have this problem

**Comment by MauDagos** [comment]
Hi @chenyiping111 thanks for your input. I think your case is solved by changing your script to:

```json
"script": {
  "lang": "painless",
  "source": "if (doc.containsKey('n_citation_ln') ) {return ( 1 + doc['n_citation_ln'].value) * _score}else{return 10*_score.intValue()}"
}
```

With that being said, your case is not the same as mine. My case is trying to access the `_score` inside the filter of a nested aggregation.

**Comment by chenyiping111** [comment]
Yes, it’s resolved by add doubleValue() after each “_score”, thank you very much.

**Comment by MauDagos** [comment]
I have also tried scripting the entire filter for the nested aggregation. No luck, I knew it was a really long shot.

```json
{
  "aggs": {
    "agg_01": {
      "nested": {"path": "nested_field_02"},
      "filter": {"script": {
        "script": {
          "lang": "painless",
          "source": """
            doc['nested_field_02.my_field'].value != null
            && _score.intValue() == 1
          """
        }
      }}
    }
  }
}
```

It really comes down to specifying the aggregation as type nested. Once that happens the `_score` is not accessible from the aggregation filter. I hope someone from the Elasticsearch team can give some feedback soon.

**Comment by MauDagos** [comment]
I've also attempted this hack where I try to save the `_score` in the script's `params`. Another long shot, no luck yet again.

```json
{
  "aggs": {
    "agg_01": {
      "nested": {"path": "nested_field_02"},
      "filter": {"script": {
        "script": {
          "lang": "painless",
          "params": {
            "docScore": {"script": {
              "lang": "painless",
              "source": "_score.intValue()"
            }}
          },
          "source": """
            doc['nested_field_02.my_field'].value != null
            && params.docScore == 1
          """
        }
      }}
    }
  }
}
```

This time there was an error returned: `class_cast_exception: Cannot convert [java.util.HashMap] to an integral value.`. It seems that `params.docScore` is a HashMap instead of an integral. I can't explain why.

**Comment by MauDagos** [comment]
I can access the `_score` in a nested `scripted_metric` aggregation. E.g.:

```json
{
  "aggs": {
    "agg_01": {
      "nested": {"path": "nested_field_02"},
      "scripted_metric": {
        "init_script": "params._agg.count = 0",
        "map_script": """
          if (doc['nested_field_02.my_field'].value != null
              && _score.intValue() == 1) {
            params._agg.count++;
          }
        """
      }
    }
  }
}
```

It's really awkward that I can access the `_score` in a `scripted_metric` but not in a, what should be, simpler `filter`.

**Comment by MauDagos** [comment]
I found issue #35985 which is almost the same bug I have. I say almost because that user wasn't trying to access the `_score` in the aggregation `"filter"`, but rather in a `"max"`.

Said user claims his query worked if he set `"size": 1`. I have tried this in ES 5.6.15, but no luck.

There's a commit linked to that issue, #36019,which appears to have fixed the bug. However I think it's only part of ES 7 or more.
